src/core/Table/TablePaginator.tsx

```diff
@@ -0,0 +1,324 @@
+    [userLocalization],
+  );
+
+  const pageListRef = React.useRef<HTMLDivElement | null>(null);
+
+  const [focusedIndex, setFocusedIndex] = React.useState<number>(currentPage);
+  React.useEffect(() => {
+    setFocusedIndex(currentPage);
+  }, [currentPage]);
+
+  const size = density != 'default' ? 'small' : undefined;
+
+  const pageButton = React.useCallback(
+    (index: number, tabIndex = index === focusedIndex ? 0 : -1) => (
+      <Button
+        key={index}
+        styleType='borderless'
+        className={cx({ 'iui-active': index === currentPage })}
+        style={{ transition: 'none' }}
+        onClick={() => onPageChange(index)}
+        aria-current={index === currentPage}
+        aria-label={localization.goToPageLabel(index + 1)}
+        tabIndex={tabIndex}
+        size={size}
+      >
+        {index + 1}
+      </Button>
+    ),
+    [focusedIndex, currentPage, localization, size, onPageChange],
+  );
+
+  const totalPagesCount = Math.ceil(totalRowsCount / pageSize);
+  const pageList = React.useMemo(
+    () =>
+      new Array(totalPagesCount)
+        .fill(null)
+        .map((_, index) => pageButton(index)),
+    [pageButton, totalPagesCount],
+  );
+  const [overflowRef, visibleCount] = useOverflow(pageList);
+
+  const onKeyDown = (event: React.KeyboardEvent<HTMLDivElement>) => {
+    // alt + arrow keys are used by browser/assistive technologies
+    if (event.altKey) {
+      return;
+    }
+
+    const focusPage = (delta: number) => {
+      const newFocusedIndex = getBoundedValue(
+        focusedIndex + delta,
+        0,
+        totalPagesCount - 1,
+      );
+
+      if (focusActivationMode === 'auto') {
+        onPageChange(newFocusedIndex);
+      } else {
+        setFocusedIndex(newFocusedIndex);
+        const buttonToFocus = Array.from(
+          pageListRef.current?.querySelectorAll('.iui-button') ?? [],
+        ).find(
+          (el) => el.textContent?.trim() === (newFocusedIndex + 1).toString(),
+        );
+        (buttonToFocus as HTMLButtonElement | undefined)?.focus();
+      }
+    };
+
+    switch (event.key) {
+      case 'ArrowRight': {
+        focusPage(+1);
+        event.preventDefault();
+        break;
+      }
+      case 'ArrowLeft': {
+        focusPage(-1);
+        event.preventDefault();
+        break;
+      }
+      case 'Enter':
+      case ' ':
+      case 'Spacebar': {
+        event.preventDefault();
+        if (focusActivationMode === 'manual') {
+          onPageChange(focusedIndex);
+        }
+        break;
+      }
+      default:
+        break;
+    }
+  };
+
+  const halfVisibleCount = Math.floor(visibleCount / 2);
+  let startPage = focusedIndex - halfVisibleCount;
+  let endPage = focusedIndex + halfVisibleCount + 1;
+  if (startPage < 0) {
+    endPage = Math.min(totalPagesCount, endPage + Math.abs(startPage));
+    startPage = 0;
+  }
+  if (endPage > totalPagesCount) {
+    startPage = Math.max(0, startPage - (endPage - totalPagesCount));
```

Comment:
What is this weird math? If values go out of bounds, shouldn't it just be `startPath = 0` and `endPage = totalPagesCount`

Reply:
Ugh, I thought I need to have some comment here but I don't know how to put that in readable English.
It helps to maintain all possible visible pages. If current page is e.g. 2 and visible count is 8, instead of showing pages from 1 to 6 (2 + (8 / 2)), I am showing from 1 to 8. I am taking the overflow part (if you can call it like that) and bringing it to the other side.

Reply:
Maybe something like this?
```suggestion
  if (startPage < 0) {
    endPage = Math.min(totalPagesCount, endPage + Math.abs(startPage)); // if no room at the beginning, show extra pages at the end
    startPage = 0;
  }
  if (endPage > totalPagesCount) {
    startPage = Math.max(0, startPage - (endPage - totalPagesCount)); // show extra pages at the beginning
```

Reply:
Added.